packages/eslint-plugin/src/rules/no-redundant-type-constituents.ts

```diff
@@ -87,107 +43,257 @@ function addToMapGroup<Key, Value>(
   }
 }
 
-function describeLiteralType(type: ts.Type): string {
-  if (type.isStringLiteral()) {
-    return JSON.stringify(type.value);
-  }
+function isNodeInsideReturnType(node: TSESTree.TSUnionType): boolean {
+  return (
+    node.parent.type === AST_NODE_TYPES.TSTypeAnnotation &&
+    isFunctionOrFunctionType(node.parent.parent)
+  );
+}
 
-  if (isTypeBigIntLiteralType(type)) {
-    return `${type.value.negative ? '-' : ''}${type.value.base10Value}n`;
-  }
+function isObjectOrIntersectionType(
+  type: ts.Type,
+): type is ts.IntersectionType | ts.ObjectType {
+  return tsutils.isObjectType(type) || tsutils.isIntersectionType(type);
+}
 
-  if (type.isLiteral()) {
-    // eslint-disable-next-line @typescript-eslint/no-base-to-string
-    return type.value.toString();
-  }
+enum TypeAssignability {
+  Equal,
+  SourceToTargetAssignable,
+  TargetToSourceAssignable,
+  NotAssignable,
+}
 
-  if (tsutils.isIntrinsicErrorType(type) && type.aliasSymbol) {
-    return type.aliasSymbol.escapedName.toString();
-  }
+function checkTypeAssignability(
+  sourceType: ts.Type,
+  targetType: ts.Type,
+  checker: ts.TypeChecker,
+): TypeAssignability {
+  const isSourceAssignableToTarget = checker.isTypeAssignableTo(
+    sourceType,
+    targetType,
+  );
+  const isTargetAssignableToSource = checker.isTypeAssignableTo(
+    targetType,
+    sourceType,
+  );
 
-  if (isTypeAnyType(type)) {
-    return 'any';
+  if (isSourceAssignableToTarget && isTargetAssignableToSource) {
+    return TypeAssignability.Equal;
   }
-
-  if (isTypeNeverType(type)) {
-    return 'never';
+  if (isSourceAssignableToTarget) {
+    return TypeAssignability.SourceToTargetAssignable;
   }
-
-  if (isTypeUnknownType(type)) {
-    return 'unknown';
+  if (isTargetAssignableToSource) {
+    return TypeAssignability.TargetToSourceAssignable;
   }
+  return TypeAssignability.NotAssignable;
+}
 
-  if (isTypeTemplateLiteralType(type)) {
-    return 'template literal type';
+function isTargetTypeRedundantInIntersection(
+  sourceType: ts.Type,
+  targetType: ts.Type,
+  checker: ts.TypeChecker,
+): boolean {
+  if (
+    tsutils.isUnionType(sourceType) &&
+    !tsutils.isIntrinsicBooleanType(sourceType)
```

Comment:
[Refactor] Removing this doesn't fail any unit tests. The subsequent `if` statement's same boolean check is needed for some unit tests, just not this one that I can tell?

```suggestion
    tsutils.isUnionType(sourceType)
```